### PR TITLE
Add missing tests to ECSDA

### PR DESCRIPTION
**Pull request by elopio**
Requires #1243.
This is one step towards #1147

# 🚀 Description

Complete the test coverage for ECRecover:
cases when version is 27, 28 and invalid.

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

**Comment by nventuro** [comment]
Can you rebase from `master`, now that #1243 has been merged? The diff is hard to read otherwise.

**Comment by elopio** [comment]
@nventuro updated.

File: test/library/ECDSA.test.js [hunk 1]
```diff
@@ -14,48 +14,101 @@ contract('ECDSA', function ([_, anyone]) {
     this.mock = await ECDSAMock.new();
```
Comment (frangio) [inline test/library/ECDSA.test.js]: While we're at it, can you rename this to `this.ecdsa` or something more descriptive?

Comment (elopio) [reply test/library/ECDSA.test.js]: done.

File: test/library/ECDSA.test.js [hunk 2]
```diff
@@ -14,48 +14,101 @@ contract('ECDSA', function ([_, anyone]) {
...
-    const signature = '0x5d99b6f7f6d1f73d1a26497f2b1c89b24c0993913f86e9a2d02cd69887d9c94f3c880358579d811b21dd1b7fd9bb01c1d81d10e69f0384e675c32b39643be89200';
-    (await this.mock.recover(TEST_MESSAGE, signature)).should.equal(signer);
-  });
+  context('recover with valid signature', function () {
+    context('with v0 signature', function () {
+      // Signature generated outside ganache with method web3.eth.sign(signer, message)
+      const signer = '0x2cc1166f6212628a0deef2b33befb2187d35b86c';
+      // eslint-disable-next-line max-len
+      const signatureWithoutVersion = '0x5d99b6f7f6d1f73d1a26497f2b1c89b24c0993913f86e9a2d02cd69887d9c94f3c880358579d811b21dd1b7fd9bb01c1d81d10e69f0384e675c32b39643be892';
+      context('with 00 as version value', function () {
```
[the remark this answers is not in the document]
Comment (elopio) [reply test/library/ECDSA.test.js]: done

File: test/library/ECDSA.test.js [hunk 3]
```diff
@@ -14,48 +14,101 @@ contract('ECDSA', function ([_, anyone]) {
...
 
-  it('recover v0', async function () {
-    // Signature generated outside ganache with method web3.eth.sign(signer, message)
-    const signer = '0x2cc1166f6212628a0deef2b33befb2187d35b86c';
-    // eslint-disable-next-line max-len
-    const signature = '0x5d99b6f7f6d1f73d1a26497f2b1c89b24c0993913f86e9a2d02cd69887d9c94f3c880358579d811b21dd1b7fd9bb01c1d81d10e69f0384e675c32b39643be89200';
-    (await this.mock.recover(TEST_MESSAGE, signature)).should.equal(signer);
-  });
+  context('recover with valid signature', function () {
+    context('with v0 signature', function () {
```
[the remark this answers is not in the document]
Comment (elopio) [reply test/library/ECDSA.test.js]: I don't think so. We check the version first, and if it's not 0/1/27/28, we return 0. It doesn't matter if the rest of the signature makes sense or not.

File: test/library/ECDSA.test.js [hunk 4]
```diff
@@ -14,48 +14,101 @@ contract('ECDSA', function ([_, anyone]) {
...
+      // Signature generated outside ganache with method web3.eth.sign(signer, message)
+      const signer = '0x2cc1166f6212628a0deef2b33befb2187d35b86c';
+      // eslint-disable-next-line max-len
+      const signatureWithoutVersion = '0x5d99b6f7f6d1f73d1a26497f2b1c89b24c0993913f86e9a2d02cd69887d9c94f3c880358579d811b21dd1b7fd9bb01c1d81d10e69f0384e675c32b39643be892';
+      context('with 00 as version value', function () {
+        it('works', async function () {
+          // Signature generated outside ganache with method web3.eth.sign(signer, message)
+          const version = '00';
+          const signature = signatureWithoutVersion + version;
+          (await this.mock.recover(TEST_MESSAGE, signature)).should.equal(signer);
```
[the remark this answers is not in the document]
Comment (elopio) [reply test/library/ECDSA.test.js]: I'm slightly against this. I think the duplication here makes every test easier to read independently. shouldRecoverSigner hides how recover works, and for tests I tend to value more the expressiveness of each one than the amount of code. I find `this.mock.recover(TEST_MESSAGE, signature)).should.equal(signer);` nice and short.

But there are good arguments for simplifying with the function, I can make the change. Your call @nventuro.

File: test/library/ECDSA.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (elopio) [reply test/library/ECDSA.test.js]: alright. I kind-of thing that's a myth that you can (and should) keep your tests totally independent from the implementation. But I'll follow your advice here. One less magic number anyway.

File: test/library/ECDSA.test.js [hunk 5]
```diff
@@ -11,56 +11,110 @@ const WRONG_MESSAGE = web3.sha3('Nope');
...
-  it('recover v1', async function () {
-    // Signature generated outside ganache with method web3.eth.sign(signer, message)
-    const signer = '0x1e318623ab09fe6de3c9b8672098464aeda9100e';
-    // eslint-disable-next-line max-len
-    const signature = '0x331fe75a821c982f9127538858900d87d3ec1f9f737338ad67cad133fa48feff48e6fa0c18abc62e42820f05943e47af3e9fbe306ce74d64094bdf1691ee53e001';
-    (await this.mock.recover(TEST_MESSAGE, signature)).should.equal(signer);
-  });
+      context('with 00 as version value', function () {
+        it('works', async function () {
+          // Signature generated outside ganache with method web3.eth.sign(signer, message)
```
[the remark this answers is not in the document]
Comment (elopio) [reply test/library/ECDSA.test.js]: Removed, thanks for catching it.

**Comment by elopio** [comment]
@frangio I think that research about v and refactoring everything to that new name once we understand what it is and how to get web3 to generate it should be a separate task.

File: test/library/ECDSA.test.js [hunk 6]
```diff
@@ -9,7 +9,7 @@ require('chai')
 const TEST_MESSAGE = web3.sha3('OpenZeppelin');
 const WRONG_MESSAGE = web3.sha3('Nope');
 
-contract('ECDSA', function ([_, anyone]) {
+contract.only('ECDSA', function ([_, anyone]) {
```
Comment (frangio) [inline test/library/ECDSA.test.js]: Oops!

Comment (elopio) [reply test/library/ECDSA.test.js]: I hate this tool :)

**Pull request merged**